### PR TITLE
chore: adopt go mod pattern and regenerate clients

### PR DESCRIPTION
**Pull request by MartinForReal**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

adopt go module to better support versioning in go project and remove vendor dir.

update go version
adopt go module
regenerate mock clients using mockgen

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by nilo19** [comment]
@MartinForReal the k/k is still using go 1.16 and keeping the vendor directory. Do we really need this update?

**Comment by MartinForReal** [comment]
k/k is built with 1.17.4 now. ref:[link](https://github.com/kubernetes/kubernetes/pull/106833). 
This is a minor change. We can leave it and decide later.

**Comment by andyzhangx** [comment]
run test first

File: go.mod [hunk 1]
```diff
@@ -1,34 +1,119 @@
...
+	golang.org/x/sys v0.0.0-20211205182925-97ca703d548d
 	k8s.io/api v0.22.4
 	k8s.io/apimachinery v0.22.4
 	k8s.io/apiserver v0.22.4
 	k8s.io/client-go v0.22.4
 	k8s.io/cloud-provider v0.22.4
 	k8s.io/component-base v0.22.4
 	k8s.io/controller-manager v0.22.4
-	k8s.io/klog/v2 v2.10.0
+	k8s.io/klog/v2 v2.9.0
```
Comment (nilo19) [inline go.mod:28]: Why downgrade this?

**Comment by nilo19** [comment]
Please update module by ./hack/update-dependencies.sh

**Comment by feiskyer** [comment]
Thanks for the PR. @MartinForReal  Could you share your steps of go mod updates? do you think anything are missing from current hack/update-dependencies.sh script?

**Comment by MartinForReal** [comment]
> Thanks for the PR. @MartinForReal Could you share your steps of go mod updates? 

Heres is how I did module update:

+ download tools from https://github.com/psampaz/go-mod-outdated
+ execute `go list -u -m -json all | go-mod-outdated -update -direct`
+ update modules which are not related to k8s
+ keep align with dependency requirements of k8s components.  (depends on the same version of component so that we don't need to manually add replace statement. )

> do you think anything are missing from current hack/update-dependencies.sh script?

I think dependable has done a great job.

File: go.mod (hunk 1 above)
Comment (MartinForReal) [reply go.mod:28]: I encounter a build error which indicates that custom logger is not compatible. So I checked the go.mod file of k8s.io/apiserver, changed the version to 2.9.0 and it worked. But I can't reproduce it now. I will revert it in new commit.

**Comment by feiskyer** [comment]
> Heres is how I did module update:
> 
> * download tools from https://github.com/psampaz/go-mod-outdated
> * execute `go list -u -m -json all | go-mod-outdated -update -direct`
> * update modules which are not related to k8s
> * keep align with dependency requirements of k8s components.  (depends on the same version of component so that we don't need to manually add replace statement. )

Then is it possible to add those steps into hack/update-dependencies.sh script?

> There is no need to maintain vendor if vendor dir is removed.

Yes, agreed. We could remove the vendor directory.

**Pull request closed without merging**